{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports and set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "TSV_FILE_PATH = '/home/projects/genomics/summary_stats/34662886-GCST90077560-EFO_0004326.h.tsv.gz'\n",
    "PATH_TO_VEP = '/home/ext_sofia_buyanova_gmail_com/vep_data'\n",
    "VEP_INPUT_FILE = 'vep_input_all'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hm_variant_id</th>\n",
       "      <th>hm_rsid</th>\n",
       "      <th>hm_chrom</th>\n",
       "      <th>hm_pos</th>\n",
       "      <th>hm_other_allele</th>\n",
       "      <th>hm_effect_allele</th>\n",
       "      <th>hm_beta</th>\n",
       "      <th>hm_odds_ratio</th>\n",
       "      <th>hm_ci_lower</th>\n",
       "      <th>hm_ci_upper</th>\n",
       "      <th>...</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>ci_lower</th>\n",
       "      <th>ci_upper</th>\n",
       "      <th>p_value</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_930158_C_T</td>\n",
       "      <td>rs1381099827</td>\n",
       "      <td>1</td>\n",
       "      <td>930158</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.192142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.905992</td>\n",
       "      <td>0.521708</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.192142</td>\n",
       "      <td>-0.905992</td>\n",
       "      <td>0.521708</td>\n",
       "      <td>0.597811</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs1381099827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_930165_G_A</td>\n",
       "      <td>rs201186828</td>\n",
       "      <td>1</td>\n",
       "      <td>930165</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.021754</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.332290</td>\n",
       "      <td>0.288782</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.021754</td>\n",
       "      <td>-0.332290</td>\n",
       "      <td>0.288782</td>\n",
       "      <td>0.890793</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs201186828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_930204_G_A</td>\n",
       "      <td>rs148711625</td>\n",
       "      <td>1</td>\n",
       "      <td>930204</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.105264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.210775</td>\n",
       "      <td>0.421303</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.105264</td>\n",
       "      <td>-0.210775</td>\n",
       "      <td>0.421303</td>\n",
       "      <td>0.513879</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.999956</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs148711625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_930215_A_G</td>\n",
       "      <td>rs903331232</td>\n",
       "      <td>1</td>\n",
       "      <td>930215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.081907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.422869</td>\n",
       "      <td>0.586683</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.081907</td>\n",
       "      <td>-0.422869</td>\n",
       "      <td>0.586683</td>\n",
       "      <td>0.750462</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.999983</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs903331232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_930245_G_A</td>\n",
       "      <td>rs146327803</td>\n",
       "      <td>1</td>\n",
       "      <td>930245</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.252748</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.192428</td>\n",
       "      <td>0.697924</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.252748</td>\n",
       "      <td>-0.192428</td>\n",
       "      <td>0.697924</td>\n",
       "      <td>0.265808</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.999978</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs146327803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  hm_variant_id       hm_rsid  hm_chrom  hm_pos hm_other_allele  \\\n",
       "0  1_930158_C_T  rs1381099827         1  930158               C   \n",
       "1  1_930165_G_A   rs201186828         1  930165               G   \n",
       "2  1_930204_G_A   rs148711625         1  930204               G   \n",
       "3  1_930215_A_G   rs903331232         1  930215               A   \n",
       "4  1_930245_G_A   rs146327803         1  930245               G   \n",
       "\n",
       "  hm_effect_allele   hm_beta  hm_odds_ratio  hm_ci_lower  hm_ci_upper  ...  \\\n",
       "0                T -0.192142            NaN    -0.905992     0.521708  ...   \n",
       "1                A -0.021754            NaN    -0.332290     0.288782  ...   \n",
       "2                A  0.105264            NaN    -0.210775     0.421303  ...   \n",
       "3                G  0.081907            NaN    -0.422869     0.586683  ...   \n",
       "4                A  0.252748            NaN    -0.192428     0.697924  ...   \n",
       "\n",
       "   other_allele  effect_allele      beta  ci_lower  ci_upper   p_value  \\\n",
       "0             C              T -0.192142 -0.905992  0.521708  0.597811   \n",
       "1             G              A -0.021754 -0.332290  0.288782  0.890793   \n",
       "2             G              A  0.105264 -0.210775  0.421303  0.513879   \n",
       "3             A              G  0.081907 -0.422869  0.586683  0.750462   \n",
       "4             G              A  0.252748 -0.192428  0.697924  0.265808   \n",
       "\n",
       "   effect_allele_frequency  standard_error  odds_ratio    variant_id  \n",
       "0                 0.000009        0.999991         NaN  rs1381099827  \n",
       "1                 0.000045        0.999955         NaN   rs201186828  \n",
       "2                 0.000044        0.999956         NaN   rs148711625  \n",
       "3                 0.000017        0.999983         NaN   rs903331232  \n",
       "4                 0.000022        0.999978         NaN   rs146327803  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(TSV_FILE_PATH, sep='\\t', compression='gzip')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>effect_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>930158</td>\n",
       "      <td>rs1381099827</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>930165</td>\n",
       "      <td>rs201186828</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>930204</td>\n",
       "      <td>rs148711625</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>930215</td>\n",
       "      <td>rs903331232</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>930245</td>\n",
       "      <td>rs146327803</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  base_pair_location    variant_id other_allele effect_allele\n",
       "0           1              930158  rs1381099827            C             T\n",
       "1           1              930165   rs201186828            G             A\n",
       "2           1              930204   rs148711625            G             A\n",
       "3           1              930215   rs903331232            A             G\n",
       "4           1              930245   rs146327803            G             A"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_for_annotation = ['chromosome', 'base_pair_location', 'variant_id', 'other_allele', 'effect_allele']\n",
    "df_to_vep = df[cols_for_annotation].copy()\n",
    "df_to_vep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>930158</td>\n",
       "      <td>rs1381099827</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>930165</td>\n",
       "      <td>rs201186828</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>930204</td>\n",
       "      <td>rs148711625</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>930215</td>\n",
       "      <td>rs903331232</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>930245</td>\n",
       "      <td>rs146327803</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # CHROM     POS           POS REF ALT\n",
       "0        1  930158  rs1381099827   C   T\n",
       "1        1  930165   rs201186828   G   A\n",
       "2        1  930204   rs148711625   G   A\n",
       "3        1  930215   rs903331232   A   G\n",
       "4        1  930245   rs146327803   G   A"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_to_vep.rename({\n",
    "    'chromosome': '# CHROM',\n",
    "    'variant_id': 'POS',\n",
    "    'base_pair_location': 'POS',\n",
    "    'other_allele': 'REF',\n",
    "    'effect_allele': 'ALT',\n",
    "}, axis='columns', inplace=True)\n",
    "df_to_vep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_to_vep.to_csv(os.path.join(PATH_TO_VEP, 'input', VEP_INPUT_FILE + '.vcf'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run VEP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1) Make sure that all is OK with permissions (`sudo chmod 777` to `vep_data` and `docker.soc`)\n",
    "2) Plugins:\n",
    "   1) CADD: download file (All possible SNVs of GRCh38/hg38 OR All gnomad release 3.0 SNVs - not all variants, without very rare mutations)\n",
    "   2) GO: not offline\n",
    "   3) EVE: also need to download smth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "! tmux new-session -d -s \"vep_all\" docker run -t -v {PATH_TO_VEP}:/opt/vep/.vep ensemblorg/ensembl-vep \\\n",
    "    ./vep --cache --format vcf --tab --force_overwrite --dir_cache /opt/vep/.vep/ --dir_plugins /opt/vep/.vep/Plugins/ \\\n",
    "        --assembly GRCh38 \\\n",
    "            --plugin Blosum62 \\\n",
    "            --plugin CADD,/opt/vep/.vep/custom/cadd.gnomad.genomes.r3.0.snv.tsv.gz \\\n",
    "            --plugin GO \\\n",
    "        --custom /opt/vep/.vep/custom/clinvar.vcf.gz,ClinVar,vcf,exact,0,CLNSIG,CLNREVSTAT,CLNDN \\\n",
    "        --input_file /opt/vep/.vep/input/{VEP_INPUT_FILE}.vcf \\\n",
    "        --output_file /opt/vep/.vep/output/{VEP_INPUT_FILE}.vep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
